substrate/frame/support/procedural/src/pallet/parse/tests/mod.rs

```diff
@@ -0,0 +1,259 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use once_cell::sync::Lazy;
+use std::{panic, sync::Mutex};
+use syn::parse_quote;
+
+/// Ensures that only one thread can modify/restore the `CARGO_MANIFEST_DIR` ENV var at a time,
+/// avoiding a race condition because `cargo test` runs tests in parallel.
+///
+/// Although this forces all tests that use [`simulate_manifest_dir`] to run sequentially with
+/// respect to each other, this is still several orders of magnitude faster than using UI
+/// tests, even if they are run in parallel.
+static MANIFEST_DIR_LOCK: Lazy<Mutex<()>> = Lazy::new(|| Mutex::new(()));
+
+/// Allows you to assert that the input expression resolves to an error whose string
+/// representation matches the specified regex literal.
+///
+/// ## Example:
+///
+/// ```
+/// use super::tasks::*;
+///
+/// assert_error_matches!(
+/// 	parse2::<TaskEnumDef>(quote! {
+/// 		#[pallet::task_enum]
+/// 		pub struct Something;
+/// 	}),
+/// 	"expected `enum`"
+/// );
+/// ```
+///
+/// More complex regular expressions are also possible (anything that could pass as a regex for
+/// use with the [`regex`] crate.):
+///
+/// ```ignore
+/// assert_error_matches!(
+/// 	parse2::<TasksDef>(quote! {
+/// 		#[pallet::tasks]
+/// 		impl<T: Config<I>, I: 'static> Pallet<T, I> {
+/// 			#[pallet::task_condition(|i| i % 2 == 0)]
+/// 			#[pallet::task_index(0)]
+/// 			pub fn foo(i: u32) -> DispatchResult {
+/// 				Ok(())
+/// 			}
+/// 		}
+/// 	}),
+/// 	r"missing `#\[pallet::task_list\(\.\.\)\]`"
+/// );
+/// ```
+///
+/// Although this is primarily intended to be used with parsing errors, this macro is general
+/// enough that it will work with any error with a reasonable [`core::fmt::Display`] impl.
+#[macro_export]
+macro_rules! assert_error_matches {
```

Comment:
The name sounds very generic to me (like it could be used for non-parsing stuff).  
Maybe something like `assert_parse_error_matches!` ?